objects/src/accounts/code.rs

```diff
@@ -0,0 +1,79 @@
+use super::{AccountError, Digest};
+use assembly::{parse_module, ModuleAst};
+use crypto::merkle::MerkleTree;
+use miden_core::Program; // TODO: we should be able to import it from the assembly crate
+
+// ACCOUNT CODE
+// ================================================================================================
+
+/// Describes public interface of an account.
+///
+/// Account's public interface consists of a set of account methods, each method being a Miden VM
+/// program. Thus, MAST root of each method commits to the underlying program. We commit to the
+/// entire account interface by building a simple Merkle tree out of all method MAST roots.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AccountCode {
+    method_tree: MerkleTree,
```

Comment:
I think I am missing something. How does a note specify which method to call?

My current understand is this:

- I create a note, this note will contain [code](https://github.com/0xPolygonMiden/miden-base/blob/2da3c850d74cede8aa61af2285d5250007907e6a/objects/src/notes/mod.rs#L21).
   - This code shall call a function of the recipients account, say `account.receive_asset` on a [P2ID](https://github.com/0xPolygonMiden/miden-base/issues/12). AFAIU the function call is defined with the [root of the MAST](https://github.com/0xPolygonMiden/miden-base/discussions/7#discussioncomment-4489315). Question: How do we know if the target account has this method? we would need a Merkle proof for the MAST, but that would require either a public registry of these proofs or communication with the target.

The following questions are:

- Do we need the sender to have this proof?
  - If so, what about changes to the account code? That would change the tree and invalidate the proof.
  - If not, how do we perform on-chain transactions?
- Or instead would we just assume the account has such a method, create the note, and let the target fill the merkle proof?

Reply:
I just realized this is only the public interface, and not the complete interface. Meaning this restricts what someone's else note may call, but does not limit what the account owner can call.

Wouldn't a name like `method_whitelist` or `public_interface` be better?

Reply:
> Question: How do we know if the target account has this method?

The recipient would need to provide this info. This can be done in a number of ways, here is a couple:
1. A QR code can include info on which method to call, or more generally, which interface an account exposes. For example, in addition to the account ID, the QR code could contain MAST root of `P2ID` script. This way, the sender will know that this is the script they need to use.
2. An account with on-chain data will include Miden assembly source of the interface. So, examining on-chain data would show which method can be called and what they do.

> If so, what about changes to the account code? That would change the tree and invalidate the proof.

If the tree changes, but there still a method with the MAST root expected by the note, the transaction can still be executed. This is because it is the recipient who needs to generate the proof that a method called by the note is a part of the account's interface. And they can do that even if the tree changes. Of course, if the method is removed entirely - then the note cannot be executed because there is nothing it can call on the account.

> Or instead would we just assume the account has such a method, create the note, and let the target fill the merkle proof?

Yes, the recipient is the one who need to prove that a method is part of the account's interface. This will actually be done implicitly by Miden VM (via transaction kernel).

> I just realized this is only the public interface, and not the complete interface.

Actually, public interface is the complete interface. There is nothing special that an account owner can do (and frequently, the concept of account owner may not make sense). There could be some methods on the interface which could be executed by some entities and not other - but that's a different story.

In general, it would be a good practice to include a "recall" clause with most notes. This protects agains:
- Incorrectly defining note script.
- Incorrectly entering recipient account number.
- Account removing a method from its interface before the note is consumed.